[mlir][debug] Support DICommonBlock.

abidh opened the pull request:
A COMMON block is a named area of memory that holds a collection of variables. Fortran subprograms may map the COMMON block memory area to a list of variables. A common block is represented in LLVM debug by DICommonBlock.

This PR adds support for this in MLIR. The changes are mostly mechanical apart from small change to access the DICompileUnit when the scope of the variable is DICommonBlock.

abidh committed:
[mlir][debug] Support DICommonBlock.

A COMMON block is a named area of memory that holds a collection of
variables. Fortran subprograms may map the COMMON block memory area to
a list of variables. A common block is represented in LLVM debug by
DICommonBlock.

This PR adds support for this in MLIR. The changes are mostly
mechanical apart from small change to access the DICompileUnit when the
scope of the variable is DICommonBlock.

Dinistro reviewed (approved):
We usually also added LLVM dialect roundtrip tests for such additions. Apart from that, this is LGTM!

gysit commented on mlir/lib/Target/LLVMIR/ModuleTranslation.cpp:
```diff
@@ -1064,19 +1064,28 @@ LogicalResult ModuleTranslation::convertGlobals() {
+      // 1. For the languages that support modules, the scope hierarchy can be
+      // variable -> DIModule -> DICompileUnit
+      //
+      // 2. For the Fortran common block variable, the scope hierarchy looks
+      // like as follows:
```
```suggestion
      // 2. For the Fortran common block variable, the scope hierarchy can be
```
nit:

gysit reviewed (approved):
LGTM modulo nits!

abidh committed:
Update mlir/lib/Target/LLVMIR/ModuleTranslation.cpp

Co-authored-by: Tobias Gysi <tobias.gysi@nextsilicon.com>

abidh committed:
Update mlir/lib/Target/LLVMIR/ModuleTranslation.cpp

Co-authored-by: Tobias Gysi <tobias.gysi@nextsilicon.com>

abidh committed:
Handle review comments.

Add a roundtrip test.

abidh commented:
> We usually also added LLVM dialect roundtrip tests for such additions. Apart from that, this is LGTM!

Thanks for the review. I was not sure where the round trip test go. But I have added one in `test/Dialect/LLVMIR/debuginfo.mlir`. Please let me know if that works.

Dinistro reviewed (approved):
LGTM, thanks for addressing the comments.
